### PR TITLE
[Merged by Bors] - Document container privileges

### LINKED ISSUE #255: Document why root is required
Basically: clean up and add to the docs:

1. We need to create two unix domain socket listeners in hostPath volumes where Kubelet can find them, so we need FS write permissions (one of them can theoretically be anywhere, one of them needs to be in a particular folder specified by the kubelet)
2. CSI tells us "put your files that you want to inject into the pod in this FS path please", we need FS permissions to create and write to that folder

[[source from Slack]](https://stackable-workspace.slack.com/archives/C03P3MUPJDT/p1679479001215299)

## Acceptance Criteria
- [ ] the need for root is explained somewhere in the secret operator docs (where exactly is up to the person picking up the ticket, and at the descretion of a reviewing person)

### PR DESCRIPTION
**Pull request by nightkr**
# Description

Fixes #255

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes
 
```[tasklist]
# Author
- [ ] Changes are OpenShift compatible
- [ ] CRD changes approved
- [ ] Helm chart can be installed and deployed operator works
- [ ] Integration tests passed (for non trivial changes)
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated
- [ ] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
```

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: docs/modules/secret-operator/pages/security.adoc [hunk 1]
```diff
@@ -0,0 +1,16 @@
+= Security
+
+== Container Privileges
```
Comment (sbernauer) [inline docs/modules/secret-operator/pages/security.adoc:3]: Not sure if this one is correct though
```suggestion
== Container privileges
```

File: docs/modules/secret-operator/pages/security.adoc [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+By default, the Secret Operator runs as privileged root containers. This is done for three reasons:
```
Comment (sbernauer) [inline docs/modules/secret-operator/pages/security.adoc:5]:
```suggestion
By default, the Secret Operator runs as a set of privileged root containers. This is done for three reasons:
```

**Review by sbernauer — APPROVED** [review]
LGTM, but I suggest @lfrancke has another look

File: docs/modules/secret-operator/pages/security.adoc (hunk 1 above)
Comment (fhennig) [reply docs/modules/secret-operator/pages/security.adoc:3]: lowercase is correct (sentence casing for titles)

File: docs/modules/secret-operator/pages/security.adoc [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+1. We need to run as root to have permission to create the Unix Domain Socket hosting the Container Storage interface (CSI)
+   driver.
```
Comment (fhennig) [inline docs/modules/secret-operator/pages/security.adoc:8]:
```suggestion
1. We need to run as root to have permission to create the Unix domain socket hosting the Container Storage interface (CSI)
   driver. The Kubelet communicates with the CSI driver over this socket.
```

I'd suggest this blip of context for the uninitiated

**Comment by fhennig** [comment]
Yeah, I'd like to have the changes in that Seb suggested, and then it LGTM too. I've [rendered it into a preview](https://deploy-preview-374--stackable-docs.netlify.app/home/nightly/secret-operator/security.html) and it shows up (I already thought it would, but better check) :ok_hand:

File: docs/modules/secret-operator/pages/security.adoc [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+2. We need to run as root to have permission to write secret material into the pods' volume paths, as directed
+   by the CSI.
+3. We need to run as a privileged container in order to be able to mount the volume folders as RAM disks, in order to avoid
+   spilling secret material onto disk.
+
+Running as root is currently a hard requirement. Secret Operator _can_ run in an unprivileged container (by passing
+`--set securityContext.privileged=false` to `helm upgrade`), but doing so is strongly discouraged, and users
```
Comment (lfrancke) [inline docs/modules/secret-operator/pages/security.adoc:15]: If anyone were not using Helm - and I'm sorry if this is a stupid question - how would it work then?
Is this purely a function of the securityContext and the operator detects whether it runs privileged or not?

Comment (nightkr) [reply docs/modules/secret-operator/pages/security.adoc:15]: It's controlled by either the CLI argument `--privileged`, or the environment variable `PRIVILEGED`.

File: docs/modules/secret-operator/pages/security.adoc [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+== Container privileges
+
+By default, the Secret Operator runs as a set of privileged root containers. This is done for three reasons:
+
+1. We need to run as root to have permission to create the Unix Domain Socket hosting the Container Storage interface (CSI)
+   driver.
+2. We need to run as root to have permission to write secret material into the pods' volume paths, as directed
+   by the CSI.
+3. We need to run as a privileged container in order to be able to mount the volume folders as RAM disks, in order to avoid
```
[the remark this answers is not in the document]
Comment (nightkr) [reply docs/modules/secret-operator/pages/security.adoc:11]: Yeah, I'd consider that a bug. We should look into switching to ramfs instead.

[the remark this answers is not in the document]
Comment (nightkr) [reply docs/modules/secret-operator/pages/security.adoc:11]: https://lwn.net/Articles/925070/ also looks interesting, though we're far away from being able to rely on that...

[the remark this answers is not in the document]
Comment (nightkr) [reply docs/modules/secret-operator/pages/security.adoc:11]: #262

**Review by lfrancke — APPROVED** [review]

**Pull request closed without merging**